Add "sync_mods" argument to state.apply/state.sls

terminalmage opened the pull request:
The docs inaccurately stated that custom module types were synced when `state.apply` is run, but this is only the case when run without SLS targets (i.e. when a highstate is run). This PR does the following:

1. Updates the docs

2. Adds a "sync" argument to provide the functionality that the docs suggested was already present

Resolves #48025.

terminalmage committed:
Fix documentation on when custom types are synced

terminalmage committed:
Add sync option to state.apply/state.sls

This allows for custom types to be synced before executing states, when
not running a highstate.

terminalmage committed:
Add test for sync argument to state.sls

terminalmage committed:
Remove redundant mocking

This mocking was added to setup_loader_modules and is no longer needed
in individual tests.

terminalmage committed:
Update docs to include references to new "sync" argument

cachedout commented on salt/modules/state.py:
```diff
@@ -599,6 +598,22 @@ def apply_(mods=None,
 
             salt '*' state.apply test localconfig=/path/to/minion.yml
+
+    sync
```
Should this be something like `sync_mods`. The first thing this makes me think of is `synchronous` but maybe that's just me?

rallytime replied:
I like that idea ^

terminalmage replied:
Yes, that does sound better, I will get that changed.

terminalmage committed:
Rename sync -> sync_mods per review suggestion

terminalmage committed:
Change 2018.3.2 to 2018.3.3

A quick bugfix release was added after 2018.3.1, so this code will not
make it in until 2018.3.3.

terminalmage commented:
`sync` renamed to `sync_mods` per recommendation from @cachedout.

Also, I updated the versionadded to bump the version number to 2018.3.3.

rallytime commented:
@terminalmage There is a related test failure with this change:

- unit.modules.test_state.StateTestCase.test_sls_sync

https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu14-n/23710/

Can you take a look?

terminalmage committed:
Update test to reflect changed argument name

terminalmage commented:
This is fixed now, I forgot to update my test to reflect the change of the argument name.

rallytime commented on salt/modules/state.py:
```diff
@@ -599,6 +598,22 @@ def apply_(mods=None,
+            :ref:`highstate <running-highstate>` automatically syncs all custom
+            module types.
+
+        .. versionadded:: 2017.7.7,2018.3.3,Fluorine
```
Ah, i just noticed this since you mentioned updating 2018.3.3 in another comment. The `2017.7.7` should be `2017.7.8`

terminalmage replied:
Done.

terminalmage committed:
Update versionchanged
